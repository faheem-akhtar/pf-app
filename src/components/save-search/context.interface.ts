import { ApiFetcherResultType } from 'api/fetcher-result-type';
import { ApiSaveSearchCreateRequestInterface } from 'api/save-search/create/request.interface';

import { SaveSearchLoadResultInterface } from './load-result-interface';

export interface SaveSearchContextInterface {
  /**
   * list of saved searches
   */
  data: SaveSearchLoadResultInterface[];

  /**
   * List of filtered saved search
   */
  filtered: SaveSearchLoadResultInterface[];

  /**
   * Create new saved search
   */
  create: (
    request: Pick<ApiSaveSearchCreateRequestInterface, 'name' | 'frequency'>
  ) => Promise<ApiFetcherResultType<SaveSearchLoadResultInterface>>;
}
