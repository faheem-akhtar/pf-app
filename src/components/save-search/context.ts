import { createContext } from 'react';
import { SaveSearchContextInterface } from './context.interface';
import { SaveSearchLoadResultInterface } from './load-result-interface';

export const SaveSearchContext = createContext<SaveSearchContextInterface>({
  data: [],
  filtered: [],
  create: () => Promise.resolve({ ok: true, data: {} as SaveSearchLoadResultInterface, headers: {} as Headers }),
});
