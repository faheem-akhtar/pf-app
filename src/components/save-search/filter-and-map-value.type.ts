export type SaveSearchFilterAndMapValueType = string | number | string[];
