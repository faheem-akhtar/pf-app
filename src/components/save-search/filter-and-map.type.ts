import { SaveSearchFilterAndMapValueType } from './filter-and-map-value.type';

export type SaveSearchFilterAndMapType = (
  value: SaveSearchFilterAndMapValueType
) => string | number | string[] | boolean | undefined;
