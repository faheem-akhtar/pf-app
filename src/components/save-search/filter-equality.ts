import { objectCompare } from 'helpers/object/compare';
import { objectReduce } from 'helpers/object/reduce';

import { SaveSearchFiltersInterface } from './filters.interface';

function stringValuesToLowerCase<T extends Partial<SaveSearchFiltersInterface>>(filters: T): T {
  return objectReduce(
    filters,
    (acc, key, value) => ({ ...acc, [key]: typeof value === 'string' ? value.toLowerCase() : value }),
    {} as T
  );
}

export function saveSearchFilterEquality(
  filters1: Partial<SaveSearchFiltersInterface>,
  filters2: Partial<SaveSearchFiltersInterface>
): boolean {
  return objectCompare(stringValuesToLowerCase(filters1), stringValuesToLowerCase(filters2));
}
