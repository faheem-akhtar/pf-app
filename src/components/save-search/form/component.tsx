import { useContext, useState } from 'react';

import { formValidatorMaxLength } from 'helpers/form/validator/max-length';
import { formValidatorRequired } from 'helpers/form/validator/required';
import { saveSearchFrequencies } from '../frequencies';
import { useFormField } from 'helpers/form/field.hook';
import { useTranslation } from 'helpers/translation/hook';

import { ButtonComponentTypeEnum } from 'library/button/component-type.enum';
import { ButtonSizeEnum } from 'library/button/size.enum';
import { ButtonTemplate } from 'library/button/template';
import { InputBaseComponent } from 'library/input/base/component';
import { SaveSearchContext } from '../context';
import { SaveSearchFrequencyEnum } from 'enums/save-search/frequency.enum';
import { SelectFieldTemplate } from 'library/select-field/template';

import styles from './save-search-form-component.module.scss';

const MAX_CHARACTERS_LIMIT = 256;

export const SaveSearchFormComponent = ({ onSuccess }: { onSuccess: () => void }): JSX.Element => {
  const { t } = useTranslation();
  const saveSearch = useContext(SaveSearchContext);
  const [generalError, setGeneralError] = useState<string>();
  const [isLoading, setIsloading] = useState(false);
  const [name, setName] = useFormField<string>('', [
    formValidatorRequired(t('save_search/name-validation-required')),
    formValidatorMaxLength(t('save_search/name-validation-character_limit'), MAX_CHARACTERS_LIMIT, true),
  ]);
  const [frequency, setFrequency] = useFormField<SaveSearchFrequencyEnum>(SaveSearchFrequencyEnum.DAILY);
  const fields = {
    id: null,
    name: '',
    formatted_filters: null,
    frequency: SaveSearchFrequencyEnum.DAILY,
  };
  const selectedFrequency = saveSearchFrequencies.find((item) => item.value === frequency.value);

  return (
    <>
      <header className={styles['save-search-form__header']}>
        <h1 className={styles['save-search-form__header-title']}>{t('save_search/form-add-title')}</h1>
      </header>
      <div className={styles['save-search-form__content']}>
        {generalError && <div className={styles['save-search-form__error--general']}>{generalError}</div>}
        {fields.formatted_filters && (
          <p className={styles['save-search-form__content-desc']}>{fields.formatted_filters}</p>
        )}
        <div className={styles['save-search-form__content-row']}>
          <label className={styles['save-search-form__label']}>{t('save_search/name-label')}</label>
          <InputBaseComponent
            floatPlaceholder={false}
            placeholder={t('save_search/name-placeholder')}
            value={name.value}
            error={!!name.error}
            onChange={(value): void => {
              setGeneralError('');
              setName(value);
            }}
          />
          {name.error && <p className={styles['save-search-form__error']}>{name.error}</p>}
        </div>
        <div className={styles['save-search-form__content-row']}>
          <label className={styles['save-search-form__label']} htmlFor='frequency'>
            {t('save_search/frequency-label')}
          </label>
          <SelectFieldTemplate
            value={selectedFrequency?.value}
            dropdownIcon
            options={saveSearchFrequencies.map((item) => ({ ...item, label: t(item.label) }))}
            onChange={(value): void => {
              setGeneralError('');
              setFrequency(value as SaveSearchFrequencyEnum);
            }}
          />
          {frequency.error && <p className={styles['save-search-form__error']}>{frequency.error}</p>}
          <p className={styles['save-search-form__info']}>{t('save_search/frequency-help_text')}</p>
        </div>
      </div>
      <footer className={styles['save-search-form__footer']}>
        <ButtonTemplate
          type='button'
          size={ButtonSizeEnum.regular}
          className={styles['save-search-form__button']}
          componentType={ButtonComponentTypeEnum.primary}
          onClick={(): void => {
            const searchName = name.value?.trim();
            if (!searchName || name.error) {
              setName(searchName);
              return;
            }
            setIsloading(true);
            saveSearch
              .create({
                name: searchName,
                frequency: frequency.value,
              })
              .then((response) => {
                setIsloading(false);
                if (response.ok) {
                  onSuccess();
                } else {
                  setGeneralError(t('general-error-message'));
                }
              });
          }}
          disabled={isLoading}
          loading={isLoading}
        >
          {t('save')}
        </ButtonTemplate>
      </footer>
    </>
  );
};
