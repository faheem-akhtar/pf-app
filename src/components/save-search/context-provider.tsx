import { FunctionComponent, useContext, useEffect, useState } from 'react';

import { apiSaveSearchCreateFetcher } from 'api/save-search/create/fetcher';
import { useApiSaveSearch } from 'api/save-search/hook';
import { FiltersContext } from 'components/filters/context';
import { UserContext } from 'components/user/context';
import { FiltersParametersEnum } from 'enums/filters/parameters.enum';

import { SaveSearchContext } from './context';
import { SaveSearchContextInterface } from './context.interface';
import { saveSearchFilterEquality } from './filter-equality';
import { saveSearchFiltersMapper } from './filters-mapper';
import { SaveSearchInterface } from './interface';
import { saveSearchTracker } from './tracker';

export const SaveSearchContextProvider: FunctionComponent = ({ children }) => {
  const user = useContext(UserContext);
  const saveSearchResponse = useApiSaveSearch();
  const saveSearchResponseData = saveSearchResponse.ok ? saveSearchResponse.data : [];
  const [searches, setSearches] = useState<SaveSearchInterface[]>([]);
  const filtersCtx = useContext(FiltersContext);
  const filters = saveSearchFiltersMapper(filtersCtx?.value || {});

  useEffect(() => {
    if (user && saveSearchResponse.ok) {
      setSearches(saveSearchResponseData);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [user, saveSearchResponse.ok]);

  const value: SaveSearchContextInterface = {
    ok: saveSearchResponse.ok,
    data: searches,
    filtered: searches.filter((item) => saveSearchFilterEquality(item.filters, filters)),
    create: ({ name, frequency }) =>
      apiSaveSearchCreateFetcher({ name, frequency, filters: filtersCtx.value }).then((response) => {
        if (response.ok) {
          setSearches([...searches, response.data]);
          saveSearchTracker.onCreateSuccess(
            filtersCtx.value[FiltersParametersEnum.locationsIds].map((location) => location.id)
          );
        }
        return response;
      }),
  };

  return <SaveSearchContext.Provider value={value}>{children}</SaveSearchContext.Provider>;
};
