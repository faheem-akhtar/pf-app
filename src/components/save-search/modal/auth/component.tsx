import React, { FunctionComponent, useContext, useEffect, useRef } from 'react';

import { AuthModalComponent } from 'components/auth/modal/component';
import { ModalComponent } from 'components/modal/component';
import { SaveSearchContext } from 'components/save-search/context';
import { saveSearchTracker } from 'components/save-search/tracker';
import { UserContext } from 'components/user/context';
import { functionNoop } from 'helpers/function/noop';
import { AuthService } from 'services/auth/service';
import { AuthSubscribeEventTypeEnum } from 'services/auth/subscribe-event-type.enum';
import { AuthSubscriberType } from 'services/auth/subscriber.type';

import { SaveSearchModalAuthPropsInterface } from './props.interface';

export const SaveSearchModalAuthComponent: FunctionComponent<SaveSearchModalAuthPropsInterface> = ({
  onCancel,
  onSuccess,
}) => {
  const user = useContext(UserContext);
  const saveSearch = useContext(SaveSearchContext);

  const openLoginRef = useRef(functionNoop);
  const closeLoginRef = useRef(functionNoop);

  useEffect(() => {
    const onUpdate: AuthSubscriberType = (newUser, meta) => {
      if (newUser && meta?.eventType === AuthSubscribeEventTypeEnum.login) {
        saveSearchTracker.onSignInSuccess();
      }
      if (newUser && meta?.eventType === AuthSubscribeEventTypeEnum.register) {
        saveSearchTracker.onSignUpSuccess();
      }
    };
    const unsubscribeAuth = AuthService.subscribe(onUpdate);

    openLoginRef.current();

    return (): void => unsubscribeAuth();
  });

  useEffect(() => {
    if (user && saveSearch.ok) {
      onSuccess();
    }
  }, [user, saveSearch.ok, onSuccess]);

  return (
    <ModalComponent openRef={openLoginRef} closeRef={closeLoginRef} overlay>
      <AuthModalComponent cancel={onCancel} />
    </ModalComponent>
  );
};
