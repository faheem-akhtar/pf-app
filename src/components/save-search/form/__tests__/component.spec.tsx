/**
 * @jest-environment jsdom
 */

import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { SaveSearchContext } from 'components/save-search/context';

import { SaveSearchFormComponent } from '../component';

describe('SaveSearchFormComponent', () => {
  it('should render the component', () => {
    expect(render(<SaveSearchFormComponent onSuccess={jest.fn()} />).container).toMatchSnapshot();
  });

  describe('name field', () => {
    it('should display error if name field is left empty', async () => {
      const { getByLabelText, getByText } = render(<SaveSearchFormComponent onSuccess={jest.fn()} />);

      userEvent.type(getByLabelText('save_search/name-label'), 'my name');

      userEvent.clear(getByLabelText('save_search/name-label'));

      expect(getByText('save_search/name-validation-required')).toBeInTheDocument();
    });

    it('should display max charactor limit error', () => {
      const { getByLabelText, getByText } = render(<SaveSearchFormComponent onSuccess={jest.fn()} />);
      const charactors: string[] = [];

      while (charactors.length < 256) {
        charactors.push([0, 1, 2, 3, 4, 5, 6, 7, 8, 9][charactors.length % 10].toString());
      }

      userEvent.type(getByLabelText('save_search/name-label'), charactors.join(''));

      expect(getByText('save_search/name-validation-character_limit')).toBeInTheDocument();
    });
  });

  describe('frequence field', () => {
    it('should select an option', () => {
      const { getByLabelText, getByRole } = render(<SaveSearchFormComponent onSuccess={jest.fn()} />);

      expect((getByRole('option', { name: 'daily' }) as HTMLOptionElement).selected).toBeTruthy();

      userEvent.selectOptions(getByLabelText('save_search/frequency-label'), '2');

      expect((getByRole('option', { name: 'weekly' }) as HTMLOptionElement).selected).toBeTruthy();

      userEvent.selectOptions(getByLabelText('save_search/frequency-label'), '0');

      expect((getByRole('option', { name: 'off' }) as HTMLOptionElement).selected).toBeTruthy();
    });
  });

  it('should not submit the form', async () => {
    const onSuccessMock = jest.fn();
    const { getByLabelText, getByRole, getByTestId } = render(<SaveSearchFormComponent onSuccess={onSuccessMock} />);

    userEvent.type(getByLabelText('save_search/name-label'), 'my name');

    userEvent.click(getByRole('button'));

    expect(getByTestId('loader-template')).toBeInTheDocument();

    await waitFor(() => expect(onSuccessMock).toHaveBeenCalledTimes(1));
  });

  it('should not submit the for if validation failed', () => {
    const onSuccessMock = jest.fn();
    const { getByRole } = render(<SaveSearchFormComponent onSuccess={onSuccessMock} />);

    userEvent.click(getByRole('button'));

    expect(onSuccessMock).not.toHaveBeenCalled();
  });

  it('should display a general error message if create request failed', async () => {
    const onSuccessMock = jest.fn();
    const saveSearchCreateMock = jest.fn().mockReturnValue(
      Promise.resolve({
        ok: false,
      })
    );
    const { getByLabelText, getByRole, findByText } = render(
      <SaveSearchContext.Provider
        value={{
          ok: true,
          data: [],
          filtered: [],
          create: saveSearchCreateMock,
        }}
      >
        <SaveSearchFormComponent onSuccess={onSuccessMock} />
      </SaveSearchContext.Provider>
    );

    userEvent.type(getByLabelText('save_search/name-label'), 'my name');

    userEvent.click(getByRole('button'));

    await findByText('general-error-message');

    expect(onSuccessMock).not.toHaveBeenCalled();
  });
});
