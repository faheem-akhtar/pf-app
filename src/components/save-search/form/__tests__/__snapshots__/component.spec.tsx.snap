// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveSearchFormComponent should render the component 1`] = `
<div>
  <header
    class="save-search-form__header"
  >
    <h1
      class="save-search-form__header-title"
    >
      save_search/form-add-title
    </h1>
  </header>
  <div
    class="save-search-form__content"
  >
    <div
      class="save-search-form__content-row"
    >
      <label
        class="save-search-form__label"
        for="save-search-name"
      >
        save_search/name-label
      </label>
      <div
        class="container"
      >
        <div
          class="content"
        >
          <input
            aria-invalid="false"
            aria-label="name"
            class="input"
            id="save-search-name"
            name="name"
            placeholder="save_search/name-placeholder"
            type="text"
            value=""
          />
        </div>
      </div>
      
    </div>
    <div
      class="save-search-form__content-row"
    >
      <label
        class="save-search-form__label"
        for="frequency"
      >
        save_search/frequency-label
      </label>
      <div
        class="container"
      >
        <div
          class="fieldWrapper"
        >
          <select
            class="field fieldWithIcon"
            id="frequency"
            name="frequency"
          >
            <option
              value="0"
            >
              off
            </option>
            <option
              value="1"
            >
              daily
            </option>
            <option
              value="2"
            >
              weekly
            </option>
          </select>
          <svg
            class="icon"
            viewBox="4.5 8 16 8"
          >
            <path
              clip-rule="evenodd"
              d="M4.64645 8.14645C4.47288 8.32001 4.4536 8.58944 4.58859 8.78431L4.64645 8.85355L11.6464 15.8536C11.82 16.0271 12.0894 16.0464 12.2843 15.9114L12.3536 15.8536L19.3536 8.85355C19.5488 8.65829 19.5488 8.34171 19.3536 8.14645C19.18 7.97288 18.9106 7.9536 18.7157 8.08859L18.6464 8.14645L12 14.793L5.35355 8.14645C5.17999 7.97288 4.91056 7.9536 4.71569 8.08859L4.64645 8.14645Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
      <p
        class="save-search-form__info"
      >
        save_search/frequency-help_text
      </p>
    </div>
  </div>
  <footer
    class="save-search-form__footer"
  >
    <button
      class="button primary regular save-search-form__button"
    >
      save
    </button>
  </footer>
</div>
`;
